src/nbl/asset/interchange/CGLTFLoader.h

```diff
@@ -0,0 +1,539 @@
+// Copyright (C) 2020 AnastaZIuk
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in Nabla.h
+
+#ifndef __NBL_ASSET_C_MESH_LOADER_GLTF__
+#define __NBL_ASSET_C_MESH_LOADER_GLTF__
+
+#include "BuildConfigOptions.h"
+
+#ifdef _NBL_COMPILE_WITH_GLTF_LOADER_
+
+#include "nbl/asset/ICPUImageView.h"
+#include "nbl/asset/interchange/IAssetLoader.h"
+#include "nbl/asset/CGLTFPipelineMetadata.h"
+
+namespace nbl
+{
+	namespace asset
+	{
+		//! glTF Loader capable of loading .gltf files
+		/*
+			glTF bridges the gap between 3D content creation tools and modern 3D applications 
+			by providing an efficient, extensible, interoperable format for the transmission and loading of 3D content.
+		*/
+		
+		class CGLTFLoader final : public IRenderpassIndependentPipelineLoader
+		{
+			protected:
+				virtual ~CGLTFLoader() {}
+
+			public:
+				CGLTFLoader(asset::IAssetManager* _m_assetMgr);
+
+				bool isALoadableFileFormat(io::IReadFile* _file) const override;
+
+				const char** getAssociatedFileExtensions() const override
+				{
+					static const char* extensions[]{ "gltf", nullptr };
+					return extensions;
+				}
+
+				uint64_t getSupportedAssetTypesBitfield() const override { return asset::IAsset::ET_MESH; }
+
+				asset::SAssetBundle loadAsset(io::IReadFile* _file, const asset::IAssetLoader::SAssetLoadParams& _params, asset::IAssetLoader::IAssetLoaderOverride* _override = nullptr, uint32_t _hierarchyLevel = 0u) override;
+
+				_NBL_STATIC_INLINE std::string getPipelineCacheKey(const E_PRIMITIVE_TOPOLOGY& primitiveType, const SVertexInputParams& vertexInputParams)
+				{
+					return "nbl/builtin/pipelines/loaders/glTF/" + std::to_string(primitiveType) + vertexInputParams.to_string();
+				}
+
+				_NBL_STATIC_INLINE std::string getSamplerCacheKey(const asset::ICPUSampler::SParams& params)
+				{
+					const std::size_t hash = std::hash<std::string_view>{}(std::string_view(reinterpret_cast<const char*>(&params), sizeof(params)));
+					return "nbl/builtin/samplers/loaders/glTF/" + std::to_string(hash);
+				}
+
+			private:
+
+				virtual void initialize() override;
+
+				using SSamplerCacheKey = std::string;
+				using SShaderDefinesCode = std::string;
+				using STextures = core::vector<std::pair<core::smart_refctd_ptr<ICPUImageView>, SSamplerCacheKey>>;
+
+				struct CGLTFHeader
+				{
+					uint32_t version;
+					std::optional<uint32_t> minVersion;
+					std::optional<std::string> generator;
+					std::optional<std::string> copyright;
```

Comment:
I dont understand the need for a `std::optional` over a string, the only time you'd ever need that is to differentiate between an empty string and a null string (nonexistent